File: python/registream-autolabel/dev/test_manual.py
# %% [markdown]
# # autolabel: manual smoke test
#
# Python analog of ``autolabel/stata/dev/test_manual.do``. Runs top-to-bottom
# in a Python interpreter, as a script, or cell-by-cell in Jupyter / VS
# Code's interactive window (each ``# %%`` starts a new cell).
#
# Prerequisites (run once in a shell before starting):
#
# ```bash
# cd ~/Github/registream-org/autolabel/python/registream-autolabel
# rm -rf ~/.registream/autolabel ~/.registream/datasets.csv  # optional clean slate
# uv sync
# ```
#
# The dev API server at ``localhost:5000`` must be running for the
# ``update_datasets`` calls to succeed; without it the bundle loads will
# fail because the cache is empty.

# %% [markdown]
# ## 0. Host override: point at the local dev API

# %%
import os
import sys

os.environ["REGISTREAM_API_HOST"] = "http://localhost:5000"

import pandas as pd

import registream.autolabel
from registream.autolabel import scope, update_datasets

# %% [markdown]
# ## 1. Download the bundle (no-op if already cached, force=True to re-fetch)

# %%
result = update_datasets("scb", "eng", force=False)
print(result)

# %% [markdown]
# ## 2. Lookups
#
# Stata shows clickable SMCL hyperlinks that drill from one command into
# the next. Jupyter / plain Python does not have that affordance, so
# instead of clicking, copy the command shown below each result and paste
# it into the next cell. Every follow-up is a real Python expression.

# %%
pd.DataFrame({"x": [1]}).rs.lookup("carb", domain="scb", lang="eng")

# %%
pd.DataFrame({"x": [1]}).rs.lookup("ssyk3", domain="scb", lang="swe")

# %%
pd.DataFrame({"x": [1]}).rs.lookup("ssyk3", domain="scb", lang="swe", detail=True)

# %% [markdown]
# ## 3. Browse scopes
#
# The module-level ``registream.autolabel.scope()`` is the Python analog
# of ``autolabel scope [<filter>]`` in Stata. At each level it returns a
# DataFrame that Jupyter renders as a table. To "click" into a row,
# copy the ``scope_level_1`` value into the next ``scope=[...]`` call.

# %%
scope(domain="scb", lang="eng")

# %%
scope(domain="scb", lang="eng", search="lisa")

# %%
scope(domain="scb", lang="eng", search="election")

# %%
scope(domain="scb", lang="swe", search="kommunfullmäktige")

# %% [markdown]
# ## 4. Waterfall: drill from a scope down to a single variable's labels

# %%
scope(domain="scb", lang="eng", search="LISA")

# %%
# Level-2 variants under LISA
scope(domain="scb", lang="eng", scope=["LISA"])

# %%
# Releases for one (scope_level_1, scope_level_2) atom
scope(
    domain="scb",
    lang="eng",
    scope=["LISA", "Individuals aged 16 and older"],
)

# %%
# Overflow: third token → release, output is the variables at that atom
scope(
    domain="scb",
    lang="eng",
    scope=["LISA", "Individuals aged 16 and older", "2005"],
)

# %%
# Variable metadata, including the alternative (STATIV) scope
pd.DataFrame({"x": [1]}).rs.lookup("kon", domain="scb", lang="eng", detail=True)

# %% [markdown]
# ## 5. Load the example dataset and auto-label
#
# Uses the shared example at ``autolabel/examples/lisa.dta`` so the
# Python and Stata dev scripts exercise the same data.

# %%
from pathlib import Path

EXAMPLES = Path(__file__).resolve().parents[3] / "examples" / "lisa.dta"
df = pd.read_stata(EXAMPLES)
df.head()

# %%
# Inference should detect LISA from the column names.
df.rs.autolabel(domain="scb", lang="eng")
print("schema_version:", df.attrs.get("schema_version"))
print("inferred scope:", df.attrs["registream"]["scope"])
df.rs.variable_labels()

# %%
df.rs.value_labels()

# %%
# Labeled view (non-mutating): codes shown with labels
df.rs.lab.head()

# %% [markdown]
# ## 6. Reload and pin the scope explicitly

# %%
df = pd.read_stata(EXAMPLES)
df.rs.autolabel(domain="scb", lang="eng", scope=["LISA"])
df.rs.variable_labels()

# %% [markdown]
# ## 7. Release pin: overflow-token syntax
#
# Passing ``scope_depth + 1`` tokens promotes the last one to
# ``release=...``. Mirrors Stata's rule at ``autolabel.ado:1499–1502``.

# %%
df = pd.read_stata(EXAMPLES)
df.rs.autolabel(
    domain="scb",
    lang="eng",
    scope=["LISA", "Individuals aged 16 and older", "2005"],
)
print("release:", df.attrs["registream"]["release"])
df.rs.variable_labels()

# %% [markdown]
# ## 8. Lookup with scope narrowing

# %%
pd.DataFrame({"x": [1]}).rs.lookup(
    "kon",
    domain="scb",
    lang="eng",
    scope=["STATIV"],
    detail=True,
)

# %% [markdown]
# ## 9. Seaborn integration: plots "just work"
#
# Importing ``registream.autolabel`` lazily wraps the common seaborn
# plotting functions so passing ``data=df`` (or ``data=df.rs.lab``) pulls
# value labels onto the categorical axes and sets axis titles from
# ``variable_labels``. Opt out with ``REGISTREAM_NO_PLOT_PATCH=1`` before
# importing.

# %%
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_stata(EXAMPLES)
df.autolabel(domain="scb", lang="eng")

categorical_cols = list(df.value_labels().keys())
first = categorical_cols[0]

fig, ax = plt.subplots(figsize=(8, 4))
sns.countplot(data=df, x=first, ax=ax)
ax.tick_params(axis="x", rotation=30)
plt.tight_layout()
# In a non-interactive script, save to disk instead of plt.show() (which
# blocks waiting for window-close). Cell-by-cell in Jupyter still renders.
if "VSCODE_PID" in os.environ or hasattr(sys, "ps1"):
    plt.show()
else:
    out = Path(__file__).parent / "test_manual_plot.png"
    fig.savefig(out, dpi=100)
    print(f"saved plot to {out}")

File: r/dev/test_manual.R
# dev/test_manual.R: R analog of autolabel/stata/dev/test_manual.do
# and autolabel/python/registream-autolabel/dev/test_manual.py.
#
# Walks the same lookup -> scope-browse -> waterfall -> autolabel
# -> pin-scope -> pin-release sequence as the Stata and Python files,
# so all three clients exercise the same flows against the same
# autolabel/examples/lisa.dta fixture.
#
# Not shipped in the package (`^dev$` is in `.Rbuildignore`).
#
# Pre-flight (run once in a shell before sourcing this file):
#
#   cd ~/Github/registream-org/autolabel/r
#   Rscript -e 'devtools::load_all()'   # editable, no install needed
#
# The dev API server at localhost:5000 must be running for the
# rs_update_datasets() call in section 1 to succeed; without it the
# bundle load will fail unless the cache at ~/.registream/autolabel/
# is already populated.
#
# Capture a log:
#
#   Rscript dev/test_manual.R 2>&1 | tee dev/test_manual.log

suppressPackageStartupMessages({
  # Load source trees (works whether or not the package is installed; the
  # installed copy may be stale relative to the working source).
  pkgload::load_all("../../registream/r", quiet = TRUE)
  pkgload::load_all(".", quiet = TRUE)
  library(haven)
})

# 0. Host override: point at the local dev API
Sys.setenv(REGISTREAM_API_HOST = "http://localhost:5000")
Sys.setenv(REGISTREAM_AUTO_APPROVE = "yes")
# Force the shared cache (matches Stata + Python). In CRAN-only mode the
# R client would prompt for shared-vs-isolated; this bypass is dev-only.
Sys.setenv(REGISTREAM_DIR = path.expand("~/.registream"))

EXAMPLES <- normalizePath(
  file.path(getwd(), "..", "examples", "lisa.dta"),
  mustWork = FALSE
)
if (!file.exists(EXAMPLES)) {
  EXAMPLES <- system.file("extdata", "lisa.dta", package = "autolabel")
}
cat(sprintf("\n=== Fixture: %s ===\n", EXAMPLES))


cat("\n=== 1. Download (or refresh) the SCB English bundle ===\n")
print(rs_update_datasets("scb", "eng", force = FALSE))


cat("\n=== 2. Lookups ===\n")
print(rs_lookup("carb", domain = "scb", lang = "eng"))
print(rs_lookup("ssyk*", domain = "scb", lang = "swe"))
print(rs_lookup("ssyk", domain = "scb", lang = "swe", detail = TRUE))


cat("\n=== 3. Scope browse ===\n")
print(scope(domain = "scb", lang = "eng"))
print(scope(domain = "scb", lang = "eng", search = "lisa"))
print(scope(domain = "scb", lang = "eng", search = "election"))
print(scope(domain = "scb", lang = "swe", search = "kommunfullmäktige"))


cat("\n=== 4. Waterfall: drill from a scope down to releases ===\n")
print(scope("LISA", domain = "scb", lang = "eng"))
print(scope("LISA", "Individuals aged 16 and older",
            domain = "scb", lang = "eng"))
# Overflow: third token promotes to release()
print(scope("LISA", "Individuals aged 16 and older", "2005",
            domain = "scb", lang = "eng"))


cat("\n=== 5. Variable-level lookup with scope narrowing ===\n")
print(rs_lookup("kon", domain = "scb", lang = "eng", detail = TRUE))


cat("\n=== 6. Load lisa.dta and auto-label (inferred scope) ===\n")
df <- as.data.frame(haven::read_dta(EXAMPLES))
cat(sprintf("Read lisa.dta: %d rows x %d cols\n", nrow(df), ncol(df)))
labelled <- autolabel(df, domain = "scb", lang = "eng")
n_var <- sum(vapply(labelled,
                    function(c) !is.null(attr(c, "label", exact = TRUE)),
                    logical(1)))
n_val <- sum(vapply(labelled,
                    function(c) inherits(c, "haven_labelled"),
                    logical(1)))
cat(sprintf("Variable labels: %d/%d, value-labelled: %d/%d\n",
            n_var, ncol(labelled), n_val, ncol(labelled)))


cat("\n=== 7. Auto-label with explicit scope pin (LISA) ===\n")
df <- as.data.frame(haven::read_dta(EXAMPLES))
labelled_pinned <- autolabel(df, domain = "scb", lang = "eng",
                             scope = "LISA")
n_var <- sum(vapply(labelled_pinned,
                    function(c) !is.null(attr(c, "label", exact = TRUE)),
                    logical(1)))
cat(sprintf("Variable labels with scope='LISA': %d/%d\n",
            n_var, ncol(labelled_pinned)))


cat("\n=== 8. Auto-label with release pin (overflow-token syntax) ===\n")
df <- as.data.frame(haven::read_dta(EXAMPLES))
labelled_release <- autolabel(
  df, domain = "scb", lang = "eng",
  scope = c("LISA", "Individuals aged 16 and older", "2005")
)
rs_attr <- attr(labelled_release, "registream", exact = TRUE)
cat(sprintf("Resolved release: %s\n",
            if (!is.null(rs_attr$release)) rs_attr$release else "<none>"))


cat("\n=== 9. Multi-domain scope sweep (parity with test_manual.do tail) ===\n")
sweeps <- list(
  c("scb",      "eng"), c("scb",      "swe"),
  c("dst",      "eng"), c("dst",      "dan"),
  c("ssb",      "eng"), c("ssb",      "nor"),
  c("sos",      "eng"), c("sos",      "swe"),
  c("fk",       "eng"), c("fk",       "swe"),
  c("hagstofa", "eng"), c("hagstofa", "isl")
)
for (pair in sweeps) {
  cat(sprintf("\n--- scope(domain='%s', lang='%s') ---\n", pair[[1]], pair[[2]]))
  out <- tryCatch(
    scope(domain = pair[[1]], lang = pair[[2]]),
    error = function(e) {
      cat(sprintf("  ERROR: %s\n", conditionMessage(e)))
      NULL
    }
  )
  if (!is.null(out)) cat(sprintf("  rows: %d\n", nrow(out)))
}


cat("\n=== 10. rs_lab on the inferred-label result ===\n")
print(utils::head(rs_lab(labelled), 5L))


cat("\n=== DONE ===\n")
